File: rust/src/bot/keyboards.rs
//! Pre-built reply keyboards. Same shapes as `bot/keyboards.py` on the
//! Python side. The teloxide adapter renders these into real
//! `KeyboardMarkup` values at the wire.

use super::Keyboard;
use crate::config::AppConfig;
use crate::models::Registration;

pub fn initial() -> Keyboard {
    Keyboard::Reply(vec![vec!["/help".into(), "/register".into()]])
}

pub fn main() -> Keyboard {
    Keyboard::Reply(vec![
        vec!["/help".into(), "/status".into()],
        vec!["/request".into()],
    ])
}

pub fn orga() -> Keyboard {
    Keyboard::Reply(vec![
        vec!["/help".into(), "/help2".into()],
        vec!["/all".into(), "/tickets".into(), "/move".into()],
        vec!["/wip".into(), "/close".into()],
    ])
}

pub fn for_user(reg: Option<&Registration>, config: &AppConfig) -> Keyboard {
    match reg {
        None => initial(),
        Some(r) if config.is_orga(&r.group_name) => orga(),
        Some(_) => main(),
    }
}

// --- /request step keyboards --------------------------------------------

pub fn categories() -> Keyboard {
    Keyboard::Reply(vec![
        vec!["Becher".into(), "Geld".into(), "/cancel".into()],
        vec!["Bier".into(), "Cocktail".into(), "Sonstiges".into()],
        vec!["Helfer".into()],
    ])
}

pub fn money() -> Keyboard {
    Keyboard::Reply(vec![
        vec!["Geld Abholen".into(), "Wechselgeld".into()],
        vec!["Freitext".into(), "/cancel".into()],
    ])
}

pub fn money_change() -> Keyboard {
    Keyboard::Reply(vec![
        vec!["Scheine".into(), "Münzen".into()],
        vec!["/cancel".into()],
    ])
}

pub fn cups() -> Keyboard {
    Keyboard::Reply(vec![
        vec!["Dreckige Abholen".into(), "/cancel".into()],
        vec!["Shotbecher".into(), "Normale Becher".into()],
    ])
}

pub fn amount() -> Keyboard {
    Keyboard::Reply(vec![
        vec!["0".into(), "Freitext".into(), "/cancel".into()],
        vec!["~10".into(), "~20".into(), "~50".into()],
    ])
}

pub fn helper() -> Keyboard {
    Keyboard::Reply(vec![
        vec!["zu viele".into(), "zu wenige".into()],
        vec!["Helfer nicht da".into(), "Liste Schichten".into()],
    ])
}

File: src/unifestbestellbot/bot/keyboards.py
"""Reply-keyboard factories. The bot uses persistent reply keyboards for
top-level navigation (initial / main / orga) and per-step prompts inside
the /request conversation. Inline keyboards live in the flows that build
them ad-hoc from data the bot just fetched."""

from aiogram.types import KeyboardButton, ReplyKeyboardMarkup

from ..config import AppConfig
from ..models import Registration


def _kb(rows: list[list[str]]) -> ReplyKeyboardMarkup:
    return ReplyKeyboardMarkup(
        keyboard=[[KeyboardButton(text=label) for label in row] for row in rows],
        resize_keyboard=True,
        is_persistent=True,
    )


INITIAL = _kb([["/help", "/register"]])
MAIN = _kb([["/help", "/status"], ["/request"]])
ORGA = _kb(
    [
        ["/help", "/help2"],
        ["/all", "/tickets", "/move"],
        ["/wip", "/close", "/self"],
    ]
)


def for_user(reg: Registration | None, config: AppConfig) -> ReplyKeyboardMarkup:
    if reg is None:
        return INITIAL
    if config.is_orga(reg.group_name):
        return ORGA
    return MAIN
